cpp/include/cuspatial/experimental/detail/pairwise_point_in_polygon.cuh

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/error.hpp>
+#include <cuspatial/experimental/detail/is_point_in_polygon.cuh>
+#include <cuspatial/traits.hpp>
+#include <cuspatial/vec_2d.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <thrust/memory.h>
+
+#include <iterator>
+#include <type_traits>
+
+namespace cuspatial {
+namespace detail {
+
+template <class Cart2dItA,
+          class Cart2dItB,
+          class OffsetIteratorA,
+          class OffsetIteratorB,
+          class OutputIt,
+          class Cart2dItADiffType = typename std::iterator_traits<Cart2dItA>::difference_type,
+          class Cart2dItBDiffType = typename std::iterator_traits<Cart2dItB>::difference_type,
+          class OffsetItADiffType = typename std::iterator_traits<OffsetIteratorA>::difference_type,
+          class OffsetItBDiffType = typename std::iterator_traits<OffsetIteratorB>::difference_type>
+__global__ void pairwise_point_in_polygon_kernel(Cart2dItA test_points_first,
```

Comment:
Since this uses trivial linear indexing, this should be a device functor instead of a kernel, and call it from a `thrust::transform()` rather than launching a kernel. This way cuSpatial doesn't have to handle grid sizing, grid stride loops, etc.

Reply:
This sounds like a big refactor. `pairwise_point_in_polygon.cuh` uses the same implementation of `is_point_in_polygon.cuh` as `point_in_polygon.cuh`.

```
__device__ inline bool is_point_in_polygon(Cart2d const& test_point,
                                           OffsetType poly_begin,
                                           OffsetType poly_end,
                                           OffsetIterator ring_offsets_first,
                                           OffsetItDiffType const& num_rings,
                                           Cart2dIt poly_points_first,
                                           Cart2dItDiffType const& num_poly_points)
                                           ```
I could be wrong, but because it requires the (multi-)polygon iterator, and rings, and points of the polygon, it doesn't seem like it would be straightforward to write a device functor to do the same thing. Like I said I could be wrong. I am assuming that you mean for me to remove the grid stride loop and instead use `thrust::transform` to do the work like:

```

  thrust::transform(rmm::exec_policy(stream),
                    test_points_first,
                    test_points_last,
                    polygon_offsets_first,
                    output,
                    [polygon_offsets_first,
                     polygon_offsets_last,
                     poly_ring_offsets_first,
                     polygon_points_first] __device__(auto const& p) {
                      is_point_in_polygon(p,
                                          polygon_offsets_first,
                                          polygon_offsets_last,
                                          poly_ring_offsets_first,
                                          num_rings,
                                          polygon_points_first,
                                          num_poly_points);
                    });
                    ```
Obviously that function doesn't work as it is written, and would require some complex thrust combinations it seems in order to combine all of the polygon components into a nested iterator that then can be passed to `thrust::transform`, unpacked, and used in `is_point_in_polygon`. Am I in the right direction here?

Reply:
Issue created to cover this: https://github.com/rapidsai/cuspatial/issues/822

Reply:
This can also benefit a lot once we have `multipolygon_range`.